File: src/main/java/org/hildan/minecraft/mining/optimizer/statistics/PatternEvaluator.java
package org.hildan.minecraft.mining.optimizer.statistics;

import org.hildan.minecraft.mining.optimizer.blocks.Sample;
import org.hildan.minecraft.mining.optimizer.ore.OreGenerator;
import org.hildan.minecraft.mining.optimizer.patterns.DiggingPattern;

/**
 * Creates a pool of Samples to evaluate digging patterns.
 */
public class PatternEvaluator {

    private final Sample[] samples;

    public PatternEvaluator(OreGenerator oreGenerator, int nbSamples, Sample baseSample) {
        samples = new Sample[nbSamples];
        for (int i = 0; i < nbSamples; i++) {
            //noinspection HardcodedLineSeparator
            System.out.printf("\r%s generated samples", i);
            samples[i] = new Sample(baseSample);
            oreGenerator.generateInto(samples[i], 5);
        }
        System.out.println();
    }

    public Statistics evaluate(DiggingPattern pattern) {
        Statistics stats = new Statistics(samples.length);
        for (Sample refSample : samples) {
            Sample sample = new Sample(refSample);
            long initialOres = sample.getOreBlocksCount();
            stats.totalOres += initialOres;

            pattern.digInto(sample);

            stats.dugBlocks += sample.getDugBlocksCount();
            stats.foundOres += initialOres - sample.getOreBlocksCount();
        }
        return stats;
    }
}

File: src/main/kotlin/org/hildan/minecraft/mining/optimizer/statistics/PatternEvaluator.kt
package org.hildan.minecraft.mining.optimizer.statistics

import org.hildan.minecraft.mining.optimizer.blocks.Sample
import org.hildan.minecraft.mining.optimizer.patterns.DiggingPattern

/**
 * Evaluates digging patterns based on the given reference samples.
 */
class PatternEvaluator(
    private val referenceSamples: List<Sample>
) {
    private val testSamples = referenceSamples.map { Sample(it) }

    private val totalOres = testSamples.map { it.oreBlocksCount }.sum().toLong()

    fun evaluate(pattern: DiggingPattern): EvaluatedPattern {
        var dugBlocks = 0L
        var foundOres = 0L
        for (i in testSamples.indices) {
            val testSample = testSamples[i]
            testSample.resetTo(referenceSamples[i])

            val initialOres = testSample.oreBlocksCount.toLong()

            pattern.digAndFollowOres(testSample)

            dugBlocks += testSample.dugBlocksCount.toLong()
            foundOres += initialOres - testSample.oreBlocksCount
        }
        return EvaluatedPattern(pattern, Statistics(foundOres, dugBlocks, totalOres))
    }
}
